pkg/odo/genericclioptions/context.go

```diff
@@ -497,15 +502,26 @@ func (o *Context) checkComponentExistsOrFail(cmp string) {
 	}
 }
 
-// ApplyIgnore will take the current ignores []string and either ignore it (if .odoignore is used)
-// or find the .gitignore file in the directory and use that instead.
+// ApplyIgnore will take the current ignores []string and append the mandatory odo-file-index.json and
+// .git ignores; or find the .odoignore/.gitignore file in the directory and use that instead.
 func ApplyIgnore(ignores *[]string, sourcePath string) (err error) {
 	if len(*ignores) == 0 {
 		rules, err := pkgUtil.GetIgnoreRulesFromDirectory(sourcePath)
 		if err != nil {
 			util.LogErrorAndExit(err, "")
 		}
 		*ignores = append(*ignores, rules...)
+	} else {
```

Comment:
So, this does fix the infinite loop of redeploying when using the `--ignore` flag, but there is an edge case here that I think we should catch.

When using a `.odoignore` or `.gitignore` file, if the `.odo/odo-file-index.json` isn't present, `odo watch` will still go into an infinite loop. 

So, what if just always add `.odo/odo-file-index.json` to the ignores list?

Reply:
Interesting, yes that is an edge case. 

We always append the index file to .gitignore when we run the indexer on `odo push`. However in your case, it would get stuck in an infinite loop.

I also found another edge case and was able to break odo watch. If I comment `#.odo/odo-file-index.json` in .gitignore and don't use `--ignore` during `odo push`, odo doesn't add it to `.gitignore` and if that happens, `odo watch` will fail again - https://github.com/openshift/odo/issues/3961 But doing it the way you suggested resolves it. So will update it.

Reply:
updated